File: advent-of-code/python/2015_17.py
from aocd import get_data

inp = get_data(day=17, year=2015)

from collections import defaultdict

ways_to_fill = defaultdict(int)

def compute_ways_to_fill(items, capacity, n_containers = 0):
  if capacity == 0:
    ways_to_fill[n_containers] += 1
    return
  if capacity < 0:
    return
  
  remaining_items = items.copy()
  for item in items:
    remaining_items.remove(item)
    compute_ways_to_fill(remaining_items, capacity - item, n_containers + 1)

compute_ways_to_fill([int(x) for x in inp.split('\n')], 150)

answer = sum(ways_to_fill.values())
answer

answer = ways_to_fill[min(ways_to_fill.keys())]
answer

File: advent-of-code/R/2015_17.R
library(tidyverse)



items <- read_lines(input) %>% parse_number()
items

ways_to_fill <- function(items, capacity) {
  if (capacity == 0) {
    return(1)
  }
  if (capacity < 0) {
    return(0)
  }
  
  result <- 0
  for (item in items) {
    items <- items[items != item | duplicated(items)]
    result <- result + ways_to_fill(
      items,
      capacity - item
    )
  }
  result
}

ways_to_fill(items, 150)

ways_to_fill_min <- function(items, capacity, n_containers = 0) {
  if (capacity == 0) {
    return(lst(n_containers, n_ways = 1))
  }
  if (capacity < 0) {
    return(lst(n_containers, n_ways = 0))
  }
  
  cur_n_ways <- 0
  cur_n_containers <- Inf
  for (item in items) {
    items <- items[items != item | duplicated(items)]
    output <- ways_to_fill_min(
      items,
      capacity - item,
      n_containers + 1
    )
    
    if (output$n_containers == cur_n_containers) {
      cur_n_ways <- cur_n_ways + output$n_ways
    } else if (output$n_containers < cur_n_containers && output$n_ways > 0) {
      cur_n_ways <- output$n_ways
      cur_n_containers <- output$n_containers
    }
  }
  lst(
    n_containers = cur_n_containers,
    n_ways = cur_n_ways,
  )
}

result <- ways_to_fill_min(items, 150)
result

answer <- result$n_ways
answer
